clock_control: nRF5x: Non-blocking 32KHz crystal oscillator startup

cvinayak opened the pull request:
Added Kconfig option and implementation to support a
non-blocking startup of 32KHz crystal oscillator.

This will reduce the time from boot to application start
while the crystal startup happens in background.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
clock_control: nRF5x: Non-blocking 32KHz crystal oscillator startup

Added Kconfig option and implementation to support a
non-blocking startup of 32KHz crystal oscillator.

This will reduce the time from boot to application start
while the crystal startup happens in background.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

pizi-nordic reviewed (changes requested):
Unfortunately this change will break Bluetooth, as clock might be not stable when connection is established.

IMHO until BLE stack is not handling clock switch-over we should still wait for stable clock. Of course this could be done in thread initializing BLE instead of  system startup which still be better than current solution.

cvinayak committed:
clock_control: nRF5x: Fix HFSTAT and LFSTAT check

Fix the implementation to correctly check the status of
HFCLK and LFCLK states with respect to the requested
sources.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Wait for stable k32src for connected roles

Added implementation to check and wait for stable 32KHz
clock source before starting connectable/directed
advertising state and initiating state.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@pizi-nordic added implementation to wait for a stable 32KHz source before connected role establishment.

pizi-nordic commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -4614,6 +4617,28 @@ static void mayfly_xtal_stop(void *params)
 	DEBUG_RADIO_CLOSE(0);
 }
 
+#define DRV_NAME CONFIG_CLOCK_CONTROL_NRF5_K32SRC_DRV_NAME
```
How this will behave on non-Nordic chips?

cvinayak replied:
Current upstream controller is nRF specific, the newer architecture in development separates vendor specifics in lower layer modules.

pizi-nordic replied:
Ok
